### DISCUSSION TITLE
node `supportPolicy` for docker image versions

### DISCUSSION THREAD

**Discussion by travi** [renovatebot/renovate#10894] [Request Help] [answered]
## Unexpected behavior

a similar scenario to https://github.com/renovatebot/renovate/discussions/10670 caught me last week. 

the specific update was:

> This PR contains the following updates:
> 
| Package | Type | Update | Change |
|---|---|---|---|
| [node](https://togithub.com/nodejs/node) | final | major | `14` -> `16` |

with a diff of the following in the `Dockerfile`:
```diff
- FROM node:14
+ FROM node:16
```

we do also use a `.nvmrc` file in the repo, so my assumption was that the dockerfile would be updated at the same time as the `.nvmrc`. since that wasn't the case, my first instinct was to explicitly set the `supportPolicy` to `lts`, since it wasn't yet, other than default behavior + `config:base`. similar to what was mentioned in https://github.com/renovatebot/renovate/discussions/10670#discussioncomment-942546, i found that the change had no impact, which made sense once i dug deeper into the docs to find that the dockerfile is not a file that is covered by the `supportPolicy`.

## Desired behavior

so, now that i had found that information and confirmed my understanding by finding https://github.com/renovatebot/renovate/discussions/10670, i tried to step back and think through what i truly wanted. i would rather not take a step to make that update more manual, like was suggested in the mentioned discussion, since my goal in the end is for the version to be in sync between the `.nvmrc` and the dockerfile. i would normally make the version in the dockerfile a variable and populate it by reading the `.nvmrc` contents, but that is not currently an option with the way our CI pipeline operates. i expect that changing that approach will not be an easy conversation, so i think i'm at the point of wanting to continue keeping the versions independent in source control, but wanting them to be updated by renovate at the same time. 

out of curiosity, is there a reason that `supportPolicy` does not cover dockerfile image versions if the image is `node`? is it just a matter of not being supported _yet_ or was there a decision made to exclude the version in that context?

## Current configurability

short of actual `supportPolicy` compatibility, is there a recommendation for configuring the docker image version that would simulate `supportPolicy` logic? if such simulation is possible, is it also possible to group that update with the `.nvmrc` update?

**Comment by viceice** [comment, marked as answer]
For docker images you need to manually assign node versioning, then you'll get only lts versions.

We can't do this by default because we won't support docker tags like `16.0.1-alpine` then.

**Reply by travi** [reply]
> For docker images you need to manually assign node versioning, then you'll get only lts versions.

i'm missing something that is probably obvious. it sounds like you're saying that there is a way to match the node versioning approach. in this particular case, we would want `lts` since it is for repos that manage packages. for our apps, we would want tighter pinning and would probably want `lts_latest`. from [docs on docker configuration](https://docs.renovatebot.com/docker/), it looks like we should also consider [digest pinning](https://docs.renovatebot.com/docker/#digest-pinning), but i'm not quite prepared for that step yet.

what i don't see in those docs is how to opt in to node versioning for the node docker image. am i overlooking docs on that approach, or is there a recommended way of accomplishing it? when you say "manually assign", do you mean to directly create rules that duplicate the rules in those categories of `supportPolicy`?

we are currently not using the alpine images, or similar, but i would expect us to want similar rules to be applied if we were.

**Reply by travi** [reply]
finally had a chance to follow up and try to get a proper config together for this. what i didnt realize from your initial answer was how literally your answer applied. i thought you were talking more generally and that there were more details that i would need and didnt know where to find them.

the config i ended up needing once i pieced together the proper ideas was pretty simple. i added the following under `packageRules`:
```json
    {
      "matchDatasources": ["docker"],
      "semanticCommitType": "ci",
      "matchPackageNames": ["node"],
      "versioning": "node"
    }
```

for others that stumble on this thread, the base concepts that were escaping me initially were the following:
* https://docs.renovatebot.com/docker/#version-compatibility
* https://docs.renovatebot.com/configuration-options/#versioning

**Reply by ypresto** [reply]
I have similar config and facing the issue exactly same as this comment:

> We can't do this by default because we won't support docker tags like `16.0.1-alpine` then.

```json
    {
      "matchManagers": ["dockerfile"],
      "matchPackageNames": ["node"],
      "versioning": "node"
    },
```

then got this

```
node	final	patch	14.17.4-alpine3.11 -> 14.17.4
```

@viceice Is there any possible workaround for this? Just ignoring and keeping suffix could solve it, but I have no idea there is such a configuration option.

**Reply by viceice** [reply]
@ypresto  You can't use node versioning if you use docker tags with suffixes.
